hudi-io/src/main/java/org/apache/hudi/io/hfile/HFileReaderImpl.java

```diff
@@ -0,0 +1,284 @@
+    HFileBlockReader blockReader = new HFileBlockReader(
+        context, stream, trailer.getLoadOnOpenDataOffset(),
+        fileSize - HFileTrailer.getTrailerSize());
+    HFileRootIndexBlock dataIndexBlock =
+        (HFileRootIndexBlock) blockReader.nextBlock(HFileBlockType.ROOT_INDEX);
+    this.dataBlockIndexEntryMap = dataIndexBlock.readBlockIndex(trailer.getDataIndexCount(), false);
+    HFileRootIndexBlock metaIndexBlock =
+        (HFileRootIndexBlock) blockReader.nextBlock(HFileBlockType.ROOT_INDEX);
+    this.metaBlockIndexEntryMap = metaIndexBlock.readBlockIndex(trailer.getMetaIndexCount(), true);
+    HFileFileInfoBlock fileInfoBlock =
+        (HFileFileInfoBlock) blockReader.nextBlock(HFileBlockType.FILE_INFO);
+    this.fileInfo = fileInfoBlock.readFileInfo();
+    this.isMetadataInitialized = true;
+  }
+
+  @Override
+  public Option<byte[]> getMetaInfo(UTF8StringKey key) throws IOException {
+    initializeMetadata();
+    byte[] bytes = fileInfo.get(key);
+    return bytes != null ? Option.of(bytes) : Option.empty();
+  }
+
+  @Override
+  public Option<ByteBuffer> getMetaBlock(String metaBlockName) throws IOException {
+    initializeMetadata();
+    BlockIndexEntry blockIndexEntry = metaBlockIndexEntryMap.get(new UTF8StringKey(metaBlockName));
+    if (blockIndexEntry == null) {
+      return Option.empty();
+    }
+    HFileBlockReader blockReader = new HFileBlockReader(
+        context, stream, blockIndexEntry.getOffset(),
+        blockIndexEntry.getOffset() + blockIndexEntry.getSize());
+    HFileMetaBlock block = (HFileMetaBlock) blockReader.nextBlock(HFileBlockType.META);
+    return Option.of(block.readContent());
+  }
+
+  @Override
+  public long getNumKeyValueEntries() {
+    try {
+      initializeMetadata();
+      return trailer.getNumKeyValueEntries();
+    } catch (IOException e) {
+      throw new RuntimeException("Cannot read HFile", e);
+    }
+  }
+
+  @Override
+  public int seekTo(Key key) throws IOException {
+    Option<KeyValue> currentKeyValue = getKeyValue();
+    if (!currentKeyValue.isPresent()) {
+      return SEEK_TO_EOF;
+    }
+    int compareCurrent = key.compareTo(currentKeyValue.get().getKey());
+    if (compareCurrent > 0) {
+      if (currentDataBlockEntry.get().getNextBlockKey().isPresent()) {
+        int comparedNextBlockFirstKey =
+            key.compareTo(currentDataBlockEntry.get().getNextBlockKey().get());
+        if (comparedNextBlockFirstKey >= 0) {
+          // Searches the block that may contain the lookup key based the starting keys of
+          // all blocks (sorted in the TreeMap of block index entries), using binary search.
+          // The result contains the greatest key less than or equal to the given key,
+          // or null if there is no such key.
+
+          Map.Entry<Key, BlockIndexEntry> floorEntry = dataBlockIndexEntryMap.floorEntry(key);
+          if (floorEntry == null) {
+            // Key smaller than the start key of the first block
```

Comment:
should the currentPos etc be reset to the first block in this case?

Reply:
Actually, the `floorEntry` should never be null.  Throwing an `IllegalStateException` here.